### ISSUE TITLE
0.10.0 Where.where().or() use case generates bad SQL

### ISSUE THREAD

**Issue by kazetsukaimiko** [ojplg/hrorm#15]
Given the following code:

```
// The goal in this example is to build a Map of Entity.id -> EntityRelative.id.
// EntityRelative.id is what is use to query related data, so I'm caching this map up front.

List<Entity> myExamples; // Has stuff
KeylessDao<EntityRelative> entityDao = getDao(EnitityRelative.class); // Magic!
Where anyOf = Where.where();
for(Entity entity : myExamples) {
  anyOf = anyOf.or("entity_id", Operator.EQUALS, entity.getId());
}
Map<Long, Long> myEntityRelativeMap = new HashMap<>();
long count = mappingDao.foldingSelect(0L, (currentCount, mapping) -> {
  myEntityRelativeMap.put(mapping.getEntityId(), mapping.getId());
  return currentCount += 1L;
}, anyOf);
return myEntityRelativeMap;
```

This generates the SQL:

`select a.id as aid, a.entity_id as aentity_id /**/ where 1=1 AND  OR a.entity_id = ? OR a.entity_id = ? ....`

AND  OR successive is malformed SQL.

**Comment by ojplg** [comment]
Hmmm .... there is a way to write this, but it is a bit inconvenient. It's one of these things that's unexpressive in Java, since there's no way to specially handle the case of one item in a list, the way you can in languages with pattern-matching, ML-derivatives, Clojure, etc.

On a completely separate note, can you use a runLongFunction() call here?

I will ponder a better error for your use-case. It's hard, since as we have discussed, generating illegal SQL is almost always going to be possible if you have a flexible API. What errors do you want to catch and when?

**Comment by kazetsukaimiko** [comment]
Before you generate your SQL from your Where objects, you need to erase "empty" conditions. I think that ought to be simple enough to fix. I don't think this should be an error scenario.

I have a workaround for right now:
```
// Start null to avoid an empty where.
Where anyOf = null;
for(Entity entity : myExamples) {
  if (anyOf == null) {
    anyOf = Where.where("entity_id", Operator.EQUALS, entity.getId());
  } else {
    anyOf = anyOf.or("entity_id", Operator.EQUALS, entity.getId());
  }
}
// .. 
// Also, use Where.where() wrapping so that I get surrounding parenthesis. 
// Otherwise I may end up with 1=1 OR condition OR condition OR condition....
 mappingDao.foldingSelect(0L, (currentCount, mapping) -> {
  myEntityRelativeMap.put(mapping.getEntityId(), mapping.getId());
  return currentCount += 1L;
}, Where.where(anyOf));

```

This generates SQL like: 
`select a.id as aid, a.entity_id as aentity_id /**/ where 1=1 AND (a.entity_id = ? OR a.entity_id = ? OR a.entity_id = ? .... )`

Note that there's another potential problem here. If you ever allow:
`select a.id as aid, a.entity_id as aentity_id /**/ where 1=1 OR /* all_my_conditions */`
... you'll select the whole table. To be safe, all your queries should generate:
`where 1=1 AND (all_my_conditions)`

**Comment by kazetsukaimiko** [comment]
As for runLongFunction, no, I cannot. I'm not just getting a count, I'm building a Map. The count is a nice-to-have.

Really, I'm butchering in a SELECT x,y,z where x IN (a,b,c,d,e,f).

I'll take a look at this as a part of the PR I'm polishing off tonight, maybe bake in some tests to catch the condition. where 1=1 and (all_my_condititions) I suspect fixes it all on its own, but there may be an empty AND condition at the start I'd have to deal with. I'm thinking some kind of Path-like normalize() method for the query generation object structure to filter out meaningless steps.
```
/path/to/some/deep/deep/deep/area/../../../but/../goes/up
// normalize() -> 
/path/to/some/deep/goes/up
```

**Comment by kazetsukaimiko** [comment]
Unrelated, but something is causing some of my fields to come back null. Not sure if it is a hrorm bug or not, but it looks like one. Will get to the bottom of it tomorrow along with a test case to reproduce if its a real issue.

**Comment by ojplg** [comment]
I put something in to throw and exception when an attempt is made to build SQL like 

`select * where OR foo=0`

I think that's the right thing to do.

The null values are more concerning. Please work up a test case for that if you can.

**Comment by kazetsukaimiko** [comment]
False alarm on the null values. I tried to create a test to reproduce but could not, so then I tried to recreate the issue in the environment. Some of the environments I have hrorm deployed in are extremely difficult to debug, I had to build some pretty elaborate scaffolding to test my code proper. It was my error- I had multiple Column definitions with the same setter (named very similarly, so I didn't notice until I watched the column definition setting the value) causing my values to be overwritten. 

I tried to think of some validation to catch this error, I'm not sure you could do it without reflection or a mocking framework. My best idea here would be, You'd make an instance of ENTITY using the supplier, iterate over every column definition, call the setter with an example value, and then ensure none of the other columns getters return the value we expect. This exception could be thrown at the .withTYPEColumn() method call, pinpointing the cause. I'm just not sure we can generate unique TYPEs to do this test with.

The downside-  what if I intend to have my setter affect multiple values? I think that's an edge case, but one worth thinking about.

**Comment by ojplg** [comment]
I have a fix for the original problem you reported with illegal SQL being generated. Simply to throw HrormException when attempting to use a logical connector with nothing.

My inclination is just to release that as 0.10.1, since it solves the original problem.

I have some further ideas about how to extend where clause building using the generic column, but I want to wait on that until some of the other things we talked about are further along.

Do you think there is another bug other than the bad SQL?

**Comment by kazetsukaimiko** [comment]
No, if that's your resolution then this one should be all set, feel free to close. The other issue only applies to user error.

**Comment by ojplg** [comment]
I don't see anything better to do for this.

Release 0.10.1 is out.
